atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -132,37 +137,46 @@ private static InetSocketAddress lookUpAddress(Map<String, InetSocketAddress> no
                 .collectToMap();
     }
 
-    // VisibleForTesting
-    public static Set<TokenRange> getMinimalSetOfRangesForTokens(
-            Metadata metadata, Set<Token> partitionKeyTokens, SortedMap<Token, TokenRange> tokenRangesByEnd) {
-        Map<Token, TokenRange> tokenRangesByStartToken = new HashMap<>();
-        for (Token token : partitionKeyTokens) {
-            TokenRange minimalTokenRange = findTokenRange(metadata, token, tokenRangesByEnd);
+    @VisibleForTesting
+    static RangeSet<LightweightOppToken> getMinimalSetOfRangesForTokens(
+            Set<LightweightOppToken> partitionKeyTokens,
+            SortedMap<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByEnd) {
+        Map<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByStartToken = new HashMap<>();
+        for (LightweightOppToken token : partitionKeyTokens) {
+            Range<LightweightOppToken> minimalTokenRange = findTokenRange(token, tokenRangesByEnd);
             tokenRangesByStartToken.merge(
-                    minimalTokenRange.getStart(), minimalTokenRange, ClusterMetadataUtils::findLatestEndingRange);
+                    LightweightOppToken.getLowerExclusive(minimalTokenRange),
+                    minimalTokenRange,
+                    ClusterMetadataUtils::findLatestEndingRange);
         }
-        return ImmutableSet.copyOf(tokenRangesByStartToken.values());
+        return tokenRangesByStartToken.values().stream().collect(toImmutableRangeSet());
     }
 
-    private static TokenRange findTokenRange(
-            Metadata metadata, Token token, SortedMap<Token, TokenRange> tokenRangesByEnd) {
+    private static Range<LightweightOppToken> findTokenRange(
+            LightweightOppToken token, SortedMap<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByEnd) {
         if (tokenRangesByEnd.containsKey(token)) {
             return tokenRangesByEnd.get(token);
         } else if (!tokenRangesByEnd.headMap(token).isEmpty()) {
-            return metadata.newTokenRange(tokenRangesByEnd.headMap(token).lastKey(), token);
+            // handle wraparound
+            LightweightOppToken lowerBound = tokenRangesByEnd.headMap(token).lastKey();
+            if (lowerBound.toString().isEmpty()) {
+                return Range.atMost(token);
+            } else {
+                return Range.openClosed(lowerBound, token);
+            }
         } else {
-            // Confirm that the first entry in the sorted map is the wraparound range
-            TokenRange firstTokenRange = tokenRangesByEnd.get(tokenRangesByEnd.firstKey());
+            // Confirm that the first entry in the sorted map is unbounded on one side
+            Range<LightweightOppToken> firstTokenRange = tokenRangesByEnd.get(tokenRangesByEnd.firstKey());
             Preconditions.checkState(
-                    firstTokenRange.isWrappedAround(),
+                    !firstTokenRange.hasUpperBound() || !firstTokenRange.hasLowerBound(),
```

Comment:
might have missed this on pass 1: if I'm reading this correctly, !hasLowerBound() should be enough?

Reply:
I actually don't think this case is hit any more.

Previously, the token range looked something like `{ (e, b], (b, c], (c, d], (d, e] }` - thus, for some `a < b`, the `headMap` call would be empty (and we would end up at this point).

We now have a token range like `{ (-inf, b],  (b, c], (c, d], (d, e], (e, +inf) }` - thus, if `a < b`, the `headMap` call now yields `(-inf, b]`.

I'll change the else case to throw IllegalStateException, because it would mean that the token ring supplied by `metadata.getTokenRanges()` in `ClusterMetadataUtils.getTokenMappingForPartitionKeys` did not give us a complete token ring.